python/pyarrow/table.pxi

```diff
@@ -1669,26 +1669,46 @@ def table(data, names=None, schema=None, metadata=None):
             "Expected pandas DataFrame, python dictionary or list of arrays")
 
 
-def concat_tables(tables):
+def concat_tables(tables, promote=False, MemoryPool memory_pool=None):
     """
-    Perform zero-copy concatenation of pyarrow.Table objects. Raises exception
-    if all of the Table schemas are not the same
+    Perform zero-copy concatenation of pyarrow.Table objects.
+
+    If promote==False, the schemas of all the Tables must be the same
+    (modulo the metadata), otherwise an exception will be raised. The result
+    Table will share the metadata with the first table.
+
+    If promote==True, columns of the same name will be concatenated.
+    They should be of the same type, or be of type NULL, in which case it will
+    be promoted to the type of other corresponding columns with null values
+    filled.  If a table is missing a particular field, null values of the
+    appropriate type will be generated to take the place of the missing field.
+    The new schema will share the metadata with the first table. Each field in
+    the new schema will share the metadata with the first table which has the
+    field defined.
 
     Parameters
     ----------
     tables : iterable of pyarrow.Table objects
-    output_name : string, default None
-      A name for the output table, if any
+    promote: bool, default False
+        if True, concatenate tables with null-filling and type promotion.
+    memory_pool : MemoryPool, default None
+        For memory allocations, if required, otherwise use default pool
     """
     cdef:
         vector[shared_ptr[CTable]] c_tables
-        shared_ptr[CTable] c_result
+        shared_ptr[CTable] c_result_table
+        CMemoryPool* pool = maybe_unbox_memory_pool(memory_pool)
         Table table
 
     for table in tables:
         c_tables.push_back(table.sp_table)
 
-    with nogil:
-        check_status(ConcatenateTables(c_tables, &c_result))
+    if promote:
+        with nogil:
```

Comment:
I would still prefer to have the whole if-else inside the a `with nogil` instead of outside. Probably needs a type annotation for `promote` as suggested above.

Reply:
That worked. Thanks.